File: backend/src/models/Source.js
import mongoose from 'mongoose';

const sourceSchema = new mongoose.Schema({
  name: {
    type: String,
    required: true,
    unique: true,
    trim: true
  },
  url: {
    type: String,
    trim: true
  },
  description: {
    type: String,
    trim: true
  },
  // 来源可信度评分
  credibilityScore: {
    type: Number,
    default: 50,
    min: 0,
    max: 100
  },
  // 来源可信度等级
  credibilityLevel: {
    type: String,
    enum: ['Very Low', 'Low', 'Medium', 'High', 'Very High'],
    default: 'Medium'
  },
  // 来源类型
  type: {
    type: String,
    enum: ['Mainstream', 'Alternative', 'Social Media', 'Blog', 'Government', 'International'],
    default: 'Mainstream'
  },
  // 政治倾向（如果适用）
  bias: {
    type: String,
    enum: ['Left', 'Center-Left', 'Center', 'Center-Right', 'Right', 'Neutral', 'Unknown'],
    default: 'Unknown'
  },
  // 审核状态
  verified: {
    type: Boolean,
    default: false
  },
  verifiedBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  },
  verifiedDate: {
    type: Date
  },
  // 关联的新闻数量
  newsCount: {
    type: Number,
    default: 0
  },
  createdAt: {
    type: Date,
    default: Date.now
  },
  updatedAt: {
    type: Date,
    default: Date.now
  }
});

// 自动更新updatedAt字段
sourceSchema.pre('save', function(next) {
  this.updatedAt = Date.now();
  next();
});

const Source = mongoose.model('Source', sourceSchema);

export default Source;

File: backend-spring/src/main/java/com/truthmoment/antifakenews/model/Source.java
package com.truthmoment.antifakenews.model;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import lombok.Data;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;

@Data
@Entity
@Table(name = "sources")
public class Source {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotBlank
    @Column(unique = true, nullable = false)
    private String name;

    private String url;

    private String description;

    @Column(nullable = false)
    private int credibilityScore = 50;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private CredibilityLevel credibilityLevel = CredibilityLevel.MEDIUM;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private SourceType type = SourceType.MAINSTREAM;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private Bias bias = Bias.UNKNOWN;

    @Column(nullable = false)
    private boolean verified = false;

    @ManyToOne
    @JoinColumn(name = "verified_by")
    private User verifiedBy;

    private LocalDateTime verifiedDate;

    @Column(nullable = false)
    private int newsCount = 0;

    @CreationTimestamp
    private LocalDateTime createdAt;

    @UpdateTimestamp
    private LocalDateTime updatedAt;

    @OneToMany(mappedBy = "source", cascade = CascadeType.ALL, orphanRemoval = true)
    private Set<News> news = new HashSet<>();

    public enum CredibilityLevel {
        VERY_LOW, LOW, MEDIUM, HIGH, VERY_HIGH
    }

    public enum SourceType {
        MAINSTREAM, ALTERNATIVE, SOCIAL_MEDIA, BLOG, GOVERNMENT, INTERNATIONAL
    }

    public enum Bias {
        LEFT, CENTER_LEFT, CENTER, CENTER_RIGHT, RIGHT, NEUTRAL, UNKNOWN
    }
}
